src/charm.py

```diff
@@ -0,0 +1,415 @@
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{PROMETHEUS_PORT}"]}],
+                }
+            ],
+        )
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="pfcp", port=8805, protocol="UDP"),
+                ServicePort(name="bess-web", port=8000),
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+            ],
+        )
+        self._kubernetes_multus = KubernetesMultusCharmLib(
+            charm=self,
+            network_attachment_definitions=[
+                NetworkAttachmentDefinition(name="access-net"),
+                NetworkAttachmentDefinition(name="core-net"),
+            ],
+            network_annotations=[
+                NetworkAnnotation(name="access-net", interface="access"),
+                NetworkAnnotation(name="core-net", interface="core"),
+            ],
+        )
+
+    def _write_bessd_config_file(self) -> None:
+        """Write the configuration file for the 5G UPF service."""
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template(f"{CONFIG_FILE_NAME}.j2")
+        content = template.render(
+            upf_hostname=self._upf_hostname,
+            mode="af_packet",
+        )
+        self._bessd_container.push(
+            path=f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content
+        )
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
+
+    @property
+    def _upf_hostname(self) -> str:
+        """Returns the UPF hostname.
+
+        Returns:
+            str: UPF Hostname
+        """
+        return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
+
+    @property
+    def _bessd_config_file_is_written(self) -> bool:
+        """Returns whether the bessd config file was written to the workload container.
+
+        Returns:
+            bool: Whether the bessd config file was written
+        """
+        if not self._bessd_container.exists(f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
+            return False
+        logger.info("Config file is written")
+        return True
+
+    @property
+    def _pfcp_agent_config_file_is_written(self) -> bool:
+        """Returns whether the pfcp agent config file was written to the workload container.
+
+        Returns:
+            bool: Whether the pfcp agent config file was written
+        """
+        if not self._pfcp_agent_container.exists(
+            f"{PFCP_AGENT_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"
+        ):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
+            return False
+        logger.info("Config file is written")
+        return True
+
+    def _on_bessd_pebble_ready(self, event: PebbleReadyEvent) -> None:
```

Comment:
Unless we require a specific type of event in those handlers, I think we should be more permissive in the type hints and just accept an `EventBase`.

Reply:
Done